benches/hashmap_benchmark.rs

```diff
@@ -0,0 +1,284 @@
+    c.bench_function("insert_struct_value_rollback", |b| {
+        b.iter_batched(
+            || prepare_struct_insert(),
+            |(ref mut map, ref mut list)| insert_struct_vec(map, list, false),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn insert_struct_value_commit(c: &mut Criterion) {
+    c.bench_function("insert_struct_value_commit", |b| {
+        b.iter_batched(
+            || prepare_struct_insert(),
+            |(ref mut map, ref mut list)| insert_struct_vec(map, list, true),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_empty_value_rollback(c: &mut Criterion) {
+    c.bench_function("remove_empty_value_rollback", |b| {
+        b.iter_batched(
+            || prepare_remove(()),
+            |(ref mut map, ref list)| remove_vec(map, list, false),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_empty_value_commit(c: &mut Criterion) {
+    c.bench_function("remove_empty_value_commit", |b| {
+        b.iter_batched(
+            || prepare_remove(()),
+            |(ref mut map, ref list)| remove_vec(map, list, true),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_struct_value_no_read_rollback(c: &mut Criterion) {
+    c.bench_function("remove_struct_value_no_read_rollback", |b| {
+        b.iter_batched(
+            || prepare_remove(Struct::default()),
+            |(ref mut map, ref list)| remove_vec(map, list, false),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_struct_value_no_read_commit(c: &mut Criterion) {
+    c.bench_function("remove_struct_value_no_read_commit", |b| {
+        b.iter_batched(
+            || prepare_remove(Struct::default()),
+            |(ref mut map, ref list)| remove_vec(map, list, true),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn search_empty_value(c: &mut Criterion) {
+    c.bench_function("search_empty_value", |b| {
+        b.iter_batched(
+            || prepare_search(()),
+            |(ref map, ref list)| search_vec(map, list),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn search_struct_value(c: &mut Criterion) {
+    c.bench_function("search_struct_value", |b| {
+        b.iter_batched(
+            || prepare_search(Struct::default()),
+            |(ref map, ref list)| search_vec(map, list),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+criterion_group!(insert, insert_empty_value_rollback, insert_empty_value_commit, insert_struct_value_rollback, insert_struct_value_commit);
+criterion_group!(remove, remove_empty_value_rollback, remove_empty_value_commit, remove_struct_value_no_read_rollback, remove_struct_value_no_read_commit);
+criterion_group!(search, search_empty_value, search_struct_value);
+criterion_main!(insert, remove, search);
+
+
+// Utility functions:
+
+fn insert_vec(map: &mut HashMap<u32, ()>, list: &Vec<u32>, commit: bool) {
```

Comment:
I think my earlier comment was that instead of insert_vec / insert_struct_vec, you have insert_vec<K, V>(map ..., list: Vec<(K, V)>, then you can re-use this utility between your insert empty and insert struct types.

Reply:
Not doing this yet, because that would break the mem::replace strategy from the comment above, but did it for prepare_insert.